src/main/java/graphql/execution/AsyncExecutionStrategy.java

```diff
@@ -45,31 +43,21 @@ public CompletableFuture<ExecutionResult> execute(ExecutionContext executionCont
         ExecutionStrategyInstrumentationContext executionStrategyCtx = ExecutionStrategyInstrumentationContext.nonNullCtx(instrumentation.beginExecutionStrategy(instrumentationParameters, executionContext.getInstrumentationState()));
 
         MergedSelectionSet fields = parameters.getFields();
-        Set<String> fieldNames = fields.keySet();
-        Async.CombinedBuilder<FieldValueInfo> futures = Async.ofExpectedSize(fields.size());
-        List<String> resolvedFields = new ArrayList<>(fieldNames.size());
-        for (String fieldName : fieldNames) {
-            MergedField currentField = fields.getSubField(fieldName);
-
-            ResultPath fieldPath = parameters.getPath().segment(mkNameForPath(currentField));
-            ExecutionStrategyParameters newParameters = parameters
-                    .transform(builder -> builder.field(currentField).path(fieldPath).parent(parameters));
-
-            resolvedFields.add(fieldName);
-            CompletableFuture<FieldValueInfo> future = resolveFieldWithInfo(executionContext, newParameters);
-            futures.add(future);
-        }
+        List<String> fieldNames = fields.getKeys();
+        Async.CombinedBuilder<FieldValueInfo> futures = resolveFieldsWithInfo(executionContext, parameters, fields, fieldNames);
         CompletableFuture<ExecutionResult> overallResult = new CompletableFuture<>();
         executionStrategyCtx.onDispatched(overallResult);
 
         futures.await().whenComplete((completeValueInfos, throwable) -> {
-            BiConsumer<List<ExecutionResult>, Throwable> handleResultsConsumer = handleResults(executionContext, resolvedFields, overallResult);
+            BiConsumer<List<ExecutionResult>, Throwable> handleResultsConsumer = handleResults(executionContext, fieldNames, overallResult);
             if (throwable != null) {
                 handleResultsConsumer.accept(null, throwable.getCause());
                 return;
             }
             Async.CombinedBuilder<ExecutionResult> executionResultFutures = Async.ofExpectedSize(completeValueInfos.size());
             for (FieldValueInfo completeValueInfo : completeValueInfos) {
+                // tweak here since most of this is NOT needed in practice since
+                // handleResults unwraps and throws away the ER objects
                 executionResultFutures.add(completeValueInfo.getFieldValue());
```

Comment:
This uses each field value as a ER but then handleResultsConsumer will throw them away anyway!

I didn't change this but I could in future work